### ISSUE TITLE
Add support for Gitbook 3

### ISSUE THREAD

**Issue by piranna** [todvora/gitbook-plugin-image-captions#8]
On gitbook v3.0.0-pre1 they have changed the API, so now the plugin totally crash. Please update it to add support for upcoming versions.

**Comment by todvora** [comment]
Hi @piranna, 
Thanks for the heads up. I'll try to make it compatible as soon, as possible.

Best regards,
Tomas

**Comment by piranna** [comment]
You are welcome :-) @sammypesse has told me today that he will add a
"legacy plugins" mode, but it will be good to have native support :-)

**Comment by todvora** [comment]
- [x] Test against multiple Gitbook versions https://travis-ci.org/todvora/gitbook-plugin-image-captions
- [x] Create Gitbook v2 and v3 compatibility adapter

**Comment by piranna** [comment]
:+1:

**Comment by SamyPesse** [comment]
@todvora I was going to post an issue about this :) (thanks @piranna )

With GitBook v3, I've improved the [documentation for plugins](http://toolchain.gitbook.com/plugins/), I've also started a section about [gitbook-tester](http://toolchain.gitbook.com/plugins/testing.html).

Let me know if some informations are missing to bring compatibility with GitBook v3

**Comment by piranna** [comment]
> @todvora I was going to post an issue about this :) (thanks @piranna )

You are welcome :-) I'm writting my thesis with GitBook, so I'm the first one interested on having it correctly working :-D

> I've also started a section about gitbook-tester.

That's awesome! This way I could easily test [my plugin](https://github.com/piranna/gitbook-plugin-printlinks)! :-D

**Comment by todvora** [comment]
Hi @piranna and @SamyPesse,
Thanks for all the information and enhanced documentation. This plugin is since release `0.4.0` [compatible](https://travis-ci.org/todvora/gitbook-plugin-image-captions/builds/126443217) with Gitbook `2.X` and `3.0.0-pre.9`. 

I am currently missing a page->html function like [gitbook.parsePage](https://github.com/todvora/gitbook-plugin-image-captions/blob/82161381edbc1e7ab0d685f188930ae8a8800bb7/src/lib/ImageCaptions.js#L201) in 2.X releases. The 3.x compatibility of image-captions is based on [parsing raw markdown](https://github.com/todvora/gitbook-plugin-image-captions/blob/82161381edbc1e7ab0d685f188930ae8a8800bb7/src/lib/ImageCaptions3.js#L34) content, which is probably not the best solution. It would be cool, if we had some simple possibility to get rendered html of pages.

Best regards,
Tomas

**Comment by SamyPesse** [comment]
Hi @todvora 

I'm currently working on GitBook v3, let me which API for plugins you'll need and I can implement it for the next release.

**Comment by todvora** [comment]
Pefect, thank you! I'm digging through the sources of Gitbook and give you know, what could be handy for me. 

One particular thing, which could be improved - speed of the build. In 3.X, comparable builds are about 3 times slower than in 2.X releases:
- [Times of 2.6 build](https://travis-ci.org/todvora/gitbook-plugin-image-captions/jobs/126443221#L340)
- [Times of 3.0.0-pre.9 build](https://travis-ci.org/todvora/gitbook-plugin-image-captions/jobs/126443222#L353)

**Comment by SamyPesse** [comment]
If you are digging thought the sources, take a look at https://github.com/GitbookIO/gitbook/tree/fixes

It's a rewrite with immutable models and a cleaner api.

I'll take a look at speed of builds.

**Comment by cycomachead** [comment]
Things seem currently broken with `3.0.0-pre.12`? It seems like changes in pre.10 lead to this error, though I haven't had a chance to experiment further.

```
TypeError: Cannot read property 'walk' of undefined
TypeError: Cannot read property 'walk' of undefined
    at ImageCaptions3.readAllImages (/Volumes/Michael/Users/Michael/Dropbox/Projects/thesis/node_modules/gitbook-plugin-image-captions/src/lib/ImageCaptions3.js:28:15)
```

I also had a problem with the version check in index.js receiving `*` as the version, causing the semver check to error. (I temporarily hacked around that one...)

**Comment by todvora** [comment]
> I also had a problem with the version check in index.js receiving \* as the version, causing the semver check to error.

This one is [already fixed](https://github.com/todvora/gitbook-plugin-image-captions/blob/08a3da0c26f8cd2e5ebf56356b3f3fe28b5fb753/src/index.js#L9). There has been a change in Gitbook config structure, that holds Gitbook version number. 

The walk issue is fixed too. So maybe we can close this issue?

**Comment by SamyPesse** [comment]
@todvora when do you expect to release the latest fixes?

**Comment by todvora** [comment]
@SamyPesse Hopefully this week. I need to resolve https://github.com/todvora/gitbook-plugin-image-captions/issues/9 and find some way, how to keep tests compatible with version 2.x and 3.x of Gitbook (mainly due to changes in levels indexing).

**Comment by todvora** [comment]
Just released `0.4.1` with full support of gitbook@3.0. Closing this issue now, feel free to reopen it or create a new one, if there is something left to fix.

Thanks everyone for your comments, support and hard work!
